File: app/src/main/java/dev/ragnarok/fenrir/api/interfaces/IPhotosApi.java
package dev.ragnarok.fenrir.api.interfaces;

import androidx.annotation.CheckResult;
import androidx.annotation.NonNull;

import java.util.Collection;
import java.util.List;

import dev.ragnarok.fenrir.api.model.AccessIdPair;
import dev.ragnarok.fenrir.api.model.IAttachmentToken;
import dev.ragnarok.fenrir.api.model.Items;
import dev.ragnarok.fenrir.api.model.VKApiComment;
import dev.ragnarok.fenrir.api.model.VKApiPhoto;
import dev.ragnarok.fenrir.api.model.VKApiPhotoAlbum;
import dev.ragnarok.fenrir.api.model.VKApiPhotoTags;
import dev.ragnarok.fenrir.api.model.VkApiPrivacy;
import dev.ragnarok.fenrir.api.model.response.DefaultCommentsResponse;
import dev.ragnarok.fenrir.api.model.response.UploadChatPhotoResponse;
import dev.ragnarok.fenrir.api.model.response.UploadOwnerPhotoResponse;
import dev.ragnarok.fenrir.api.model.server.VkApiChatPhotoUploadServer;
import dev.ragnarok.fenrir.api.model.server.VkApiOwnerPhotoUploadServer;
import dev.ragnarok.fenrir.api.model.server.VkApiPhotoMessageServer;
import dev.ragnarok.fenrir.api.model.server.VkApiUploadServer;
import dev.ragnarok.fenrir.api.model.server.VkApiWallUploadServer;
import io.reactivex.rxjava3.core.Single;

public interface IPhotosApi {

    @CheckResult
    Single<Boolean> deleteAlbum(int albumId, Integer groupId);

    @CheckResult
    Single<Boolean> restore(Integer ownerId, int photoId);

    @CheckResult
    Single<Boolean> delete(Integer ownerId, int photoId);

    @CheckResult
    Single<Boolean> deleteComment(Integer ownerId, int commentId);

    @CheckResult
    Single<Boolean> restoreComment(Integer ownerId, int commentId);

    @CheckResult
    Single<Boolean> editComment(Integer ownerId, int commentId, String message,
                                Collection<IAttachmentToken> attachments);

    @CheckResult
    Single<VKApiPhotoAlbum> createAlbum(String title, Integer groupId, String description,
                                        VkApiPrivacy privacyView, VkApiPrivacy privacyComment,
                                        Boolean uploadByAdminsOnly, Boolean commentsDisabled);

    @CheckResult
    Single<Boolean> editAlbum(int albumId, String title, String description, Integer ownerId,
                              VkApiPrivacy privacyView, VkApiPrivacy privacyComment,
                              Boolean uploadByAdminsOnly, Boolean commentsDisabled);

    @CheckResult
    Single<Integer> copy(int ownerId, int photoId, String accessKey);

    @CheckResult
    Single<Integer> createComment(Integer ownerId, int photoId, Boolean fromGroup, String message,
                                  Integer replyToComment, Collection<IAttachmentToken> attachments,
                                  Integer stickerId, String accessKey, Integer generatedUniqueId);

    @CheckResult
    Single<DefaultCommentsResponse> getComments(Integer ownerId, int photoId, Boolean needLikes,
                                                Integer startCommentId, Integer offset, Integer count, String sort,
                                                String accessKey, Boolean extended, String fields);

    @CheckResult
    Single<List<VKApiPhoto>> getById(@NonNull Collection<AccessIdPair> ids);

    @CheckResult
    Single<VkApiUploadServer> getUploadServer(int albumId, Integer groupId);


    @CheckResult
    Single<UploadOwnerPhotoResponse> saveOwnerPhoto(String server, String hash, String photo);

    @CheckResult
    Single<VkApiOwnerPhotoUploadServer> getOwnerPhotoUploadServer(Integer ownerId);

    @CheckResult
    Single<VkApiChatPhotoUploadServer> getChatUploadServer(Integer chat_id);

    @CheckResult
    Single<UploadChatPhotoResponse> setChatPhoto(String file);

    @CheckResult
    Single<List<VKApiPhoto>> saveWallPhoto(Integer userId, Integer groupId, String photo, int server,
                                           String hash, Double latitude, Double longitude, String caption);

    @CheckResult
    Single<VkApiWallUploadServer> getWallUploadServer(Integer groupId);

    @CheckResult
    Single<List<VKApiPhoto>> save(int albumId, Integer groupId, int server, String photosList, String hash,
                                  Double latitude, Double longitude, String caption);

    @CheckResult
    Single<Items<VKApiPhoto>> get(Integer ownerId, String albumId, Collection<Integer> photoIds, Boolean rev,
                                  Integer offset, Integer count);

    @CheckResult
    Single<Items<VKApiPhoto>> getUsersPhoto(Integer ownerId, Integer extended, Integer sort, Integer offset, Integer count);

    @CheckResult
    Single<Items<VKApiPhoto>> getAll(Integer ownerId, Integer extended, Integer photo_sizes, Integer offset, Integer count);

    @CheckResult
    Single<VkApiPhotoMessageServer> getMessagesUploadServer();

    @CheckResult
    Single<List<VKApiPhoto>> saveMessagesPhoto(Integer server, String photo, String hash);

    @CheckResult
    Single<Items<VKApiPhotoAlbum>> getAlbums(Integer ownerId, Collection<Integer> albumIds, Integer offset,
                                             Integer count, Boolean needSystem, Boolean needCovers);

    @CheckResult
    Single<List<VKApiPhotoTags>> getTags(Integer ownerId, Integer photo_id, String access_key);

    @CheckResult
    Single<Items<VKApiComment>> getAllComments(Integer ownerId, Integer album_id, Integer need_likes, Integer offset, Integer count);

    @CheckResult
    Single<Items<VKApiPhoto>> search(String q, Double lat_gps, Double long_gps, Integer sort, Integer radius, Long startTime, Long endTime, Integer offset, Integer count);

}

File: app/src/main/java/dev/ragnarok/fenrir/api/interfaces/IPhotosApi.kt
package dev.ragnarok.fenrir.api.interfaces

import androidx.annotation.CheckResult
import dev.ragnarok.fenrir.api.model.*
import dev.ragnarok.fenrir.api.model.response.DefaultCommentsResponse
import dev.ragnarok.fenrir.api.model.response.UploadChatPhotoResponse
import dev.ragnarok.fenrir.api.model.response.UploadOwnerPhotoResponse
import dev.ragnarok.fenrir.api.model.server.*
import io.reactivex.rxjava3.core.Single

interface IPhotosApi {
    @CheckResult
    fun deleteAlbum(albumId: Int, groupId: Int?): Single<Boolean>

    @CheckResult
    fun restore(ownerId: Int?, photoId: Int): Single<Boolean>

    @CheckResult
    fun delete(ownerId: Int?, photoId: Int): Single<Boolean>

    @CheckResult
    fun deleteComment(ownerId: Int?, commentId: Int): Single<Boolean>

    @CheckResult
    fun restoreComment(ownerId: Int?, commentId: Int): Single<Boolean>

    @CheckResult
    fun editComment(
        ownerId: Int?, commentId: Int, message: String?,
        attachments: Collection<IAttachmentToken>?
    ): Single<Boolean>

    @CheckResult
    fun createAlbum(
        title: String?, groupId: Int?, description: String?,
        privacyView: VkApiPrivacy?, privacyComment: VkApiPrivacy?,
        uploadByAdminsOnly: Boolean?, commentsDisabled: Boolean?
    ): Single<VKApiPhotoAlbum>

    @CheckResult
    fun editAlbum(
        albumId: Int, title: String?, description: String?, ownerId: Int?,
        privacyView: VkApiPrivacy?, privacyComment: VkApiPrivacy?,
        uploadByAdminsOnly: Boolean?, commentsDisabled: Boolean?
    ): Single<Boolean>

    @CheckResult
    fun copy(ownerId: Int, photoId: Int, accessKey: String?): Single<Int>

    @CheckResult
    fun createComment(
        ownerId: Int?, photoId: Int, fromGroup: Boolean?, message: String?,
        replyToComment: Int?, attachments: Collection<IAttachmentToken>?,
        stickerId: Int?, accessKey: String?, generatedUniqueId: Int?
    ): Single<Int>

    @CheckResult
    fun getComments(
        ownerId: Int?, photoId: Int, needLikes: Boolean?,
        startCommentId: Int?, offset: Int?, count: Int?, sort: String?,
        accessKey: String?, extended: Boolean?, fields: String?
    ): Single<DefaultCommentsResponse>

    @CheckResult
    fun getById(ids: Collection<AccessIdPair>): Single<List<VKApiPhoto>>

    @CheckResult
    fun getUploadServer(albumId: Int, groupId: Int?): Single<VkApiUploadServer>

    @CheckResult
    fun saveOwnerPhoto(
        server: String?,
        hash: String?,
        photo: String?
    ): Single<UploadOwnerPhotoResponse>

    @CheckResult
    fun getOwnerPhotoUploadServer(ownerId: Int?): Single<VkApiOwnerPhotoUploadServer>

    @CheckResult
    fun getChatUploadServer(chat_id: Int?): Single<VkApiChatPhotoUploadServer>

    @CheckResult
    fun setChatPhoto(file: String?): Single<UploadChatPhotoResponse>

    @CheckResult
    fun saveWallPhoto(
        userId: Int?, groupId: Int?, photo: String?, server: Int,
        hash: String?, latitude: Double?, longitude: Double?, caption: String?
    ): Single<List<VKApiPhoto>>

    @CheckResult
    fun getWallUploadServer(groupId: Int?): Single<VkApiWallUploadServer>

    @CheckResult
    fun save(
        albumId: Int, groupId: Int?, server: Int, photosList: String?, hash: String?,
        latitude: Double?, longitude: Double?, caption: String?
    ): Single<List<VKApiPhoto>>

    @CheckResult
    operator fun get(
        ownerId: Int?, albumId: String?, photoIds: Collection<Int?>?, rev: Boolean?,
        offset: Int?, count: Int?
    ): Single<Items<VKApiPhoto>>

    @CheckResult
    fun getUsersPhoto(
        ownerId: Int?,
        extended: Int?,
        sort: Int?,
        offset: Int?,
        count: Int?
    ): Single<Items<VKApiPhoto>>

    @CheckResult
    fun getAll(
        ownerId: Int?,
        extended: Int?,
        photo_sizes: Int?,
        offset: Int?,
        count: Int?
    ): Single<Items<VKApiPhoto>>

    @get:CheckResult
    val messagesUploadServer: Single<VkApiPhotoMessageServer>

    @CheckResult
    fun saveMessagesPhoto(server: Int?, photo: String?, hash: String?): Single<List<VKApiPhoto>>

    @CheckResult
    fun getAlbums(
        ownerId: Int?, albumIds: Collection<Int?>?, offset: Int?,
        count: Int?, needSystem: Boolean?, needCovers: Boolean?
    ): Single<Items<VKApiPhotoAlbum>>

    @CheckResult
    fun getTags(ownerId: Int?, photo_id: Int?, access_key: String?): Single<List<VKApiPhotoTags>>

    @CheckResult
    fun getAllComments(
        ownerId: Int?,
        album_id: Int?,
        need_likes: Int?,
        offset: Int?,
        count: Int?
    ): Single<Items<VKApiComment>>

    @CheckResult
    fun search(
        q: String?,
        lat_gps: Double?,
        long_gps: Double?,
        sort: Int?,
        radius: Int?,
        startTime: Long?,
        endTime: Long?,
        offset: Int?,
        count: Int?
    ): Single<Items<VKApiPhoto>>
}